site/source/tutorials/005_form_widgets/index.md

```diff
@@ -1,26 +1,159 @@
+
+{% include_codefile 'demo/finished/biz-e-corp/src/widgets/App.ts' lines:73-110 %}
+
+Most of the properties are being populated by the `formData` object that is being retrieved from the `App`'s state with the only exceptions being the event handlers.
+
+The `onChange` handler is calling the `App`'s `setState` method, passing in all of the existing state (via the `...this.state.formData` statement), and appending the new state data coming in from the event (via `[field]: value`). After the event is processed, Dojo 2 will automatically detect that the state has changed and call the `invalidate` method, causing the page to be refreshed with the new data.
+
+The `onBlur` handler is almost the same as the `onChange` handler, except that it is making a call to the `_validateWorkerData` method and the result to update the `state`'s `formData`. Add that method to the `App` class as follows:
+
+{% include_codefile 'demo/finished/biz-e-corp/src/widgets/App.ts' lines:58-71 %}
+
+This method provides some simple validation rules for each field. It ensures that values have been provided for the `firstName` and `lastName` fields and that the `email` field contains a valid email address.
+
+The final event handler is the `onSubmit` handler which calls the `_addWorker` method. Add that method in next:
+
+{% include_codefile 'demo/finished/biz-e-corp/src/widgets/App.ts' lines:34-56 %}
+
+This method retrieves the `formData` object from the `App`'s `state` and, after confirming that there are no validation errors, creates a new entry in the `_workerData` array. Next, it resets the `formData` object to `undefined`, clearing the form. Finally, it calls the `invalidate` method to cause the page to be re-rendered to reflect these changes.
+
+## Final checks
+
+With the `WidgetForm` in place and the `App` setup to handle it, let's try it out. First test the [happy path](https://en.wikipedia.org/wiki/Happy_path) by providing the expected values to the form. Fill values for the fields, for example: "Suzette McNamara (smcnamara359@email.com)" and click the `Save` button. As expected, the form is cleared and a new `Worker` widget is added to the page. Clicking on that widget shows the detailed information of the card where we find that the first name, last name, and email values have been properly rendered.
+
+Now, let's try the validation logic. The "Given name" and "Family name" fields are simply checking for a non-empty value so the validation logic can be tested by focusing on each element and then hitting the `tab` key to move to the next form field. The invalid values cause the widgets' borders to change to a red color. Additionally, inspecting the DOM shows that the `aria-invalid` attribute has been set to `true`, allowing users that are using assistive technologies to be aware of the validation errors. Finally, the "Email address" field can be tested by providing a blank value or an invalid value such as "notanemail".
```

Comment:
Not sure I like that we're now putting things in quotes, can we just emphasize them in bold or something?